Client config builder

pileks opened the pull request:
Works towards resolving #1122

1st implementation of Client Config Builder. Very basic, follows the config logic that’s already present within the Client.

Future work:

- Set up config bundles for various use-cases (Ethereum/ENS, IPFS etc.)
- Create `addX` (`addPlugin`, `addEnv`, etc.) methods
- Consider & implement different config add/override behavior
  - Currently, the 1st config for a given plugin/env/etc. is used, making stuff like chaining `.add()` calls behave unintuitively
  - My personal preference for config builders is to have whatever comes last be the value used within the config
  - Possible solution - have two base methods: `add` and `set`, with their respective `addX` and `setX` for each config section (plugins, envs, etc.)
    - `add` would append, while `set` would overwrite
    - behavior would need to be examined and determined per config section (what does it mean to `add` an `interfaceImplementation` vs `add` a `plugin`, for example?
- Load config from a file? (JSON, TOML?)

Comments/thoughts very welcome!

pileks committed:
client config builder init

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
config defaults, change config builder api to strings

pileks committed:
move sanitization, add test class that needs to be removed WIP

pileks committed:
Remove unused test class and references

pileks committed:
small refactor, update tests

pileks committed:
move pluginwrapper and wasmwrapper into core

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
lint

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
fix wrong order of operations for config in _setContext

pileks committed:
Fix wrong package versions

pileks committed:
remove erroneous resolver plugin config

pileks committed:
fix setContext behavior to reflect current intended logic

pileks committed:
lint

pileks committed:
lint

pileks committed:
fix test code to user client-config-builder

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

cbrzn committed:
chore: fix conflicts

cbrzn commented on packages/js/client-config/README.md:
```diff
@@ -0,0 +1 @@
+TODO
```
Let's add a some description here

pileks replied:
Added a slightly modified version of the description found in this hackmd I'm setting up:
https://hackmd.io/0rDyONfoT4mw2HxTYtjAxA

Hope it looks good!

cbrzn reviewed (dismissed):
Thanks for pushing this forward 🙏 

LGTM 🍏

pileks committed:
Add client config builder readme

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
client config builder - addPlugin method

pileks committed:
client config builder - removePlugin method

pileks committed:
client config builder - add env, additional tests

pileks committed:
client config builder - removeEnv

pileks committed:
client config builder - setEnv

pileks committed:
client config builder - addInterfaceImplementation

pileks committed:
client config builder - addInterfaceImplementations

pileks committed:
client config builder - removeInterfaceImplementation

pileks committed:
client config builder - addUriRedirect

pileks committed:
client config builder - remove uri redirect

pileks committed:
client config builder - setUriResolvers

pileks committed:
client config builder - addUriResolver

pileks committed:
client config builder - remove old sanitization functions

pileks committed:
client config builder - add proper test for addDefaults

cbrzn committed:
chore(client-config): add dependencies in readme

cbrzn committed:
chore: fix conflicts

cbrzn reviewed (dismissed):
I went ahead and fixed the conflicts. Once again thanks for this @pileks - Amazing job :-D

pileks committed:
align client ctor with config builder logic, fix up intended breaks in test cases

pileks committed:
Merge remote-tracking branch 'origin/pileks-client-config-builder-origin' into pileks-client-config-builder-methods

pileks commented:
> I went ahead and fixed the conflicts. Once again thanks for this @pileks - Amazing job :-D

Thanks @cbrzn! I've updated another related PR with the merge, you might want to check it out: https://github.com/polywrap/monorepo/pull/1135

pileks committed:
remove unnecessary check

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
post-merge fixes

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
update package.lock in client config builder

pileks committed:
Merge remote-tracking branch 'origin/pileks-client-config-builder-origin' into pileks-client-config-builder-methods

pileks committed:
fix package.json in core

pileks committed:
Merge remote-tracking branch 'origin/origin-dev' into pileks-client-config-builder-origin

pileks committed:
Merge remote-tracking branch 'origin/pileks-client-config-builder-origin' into pileks-client-config-builder-methods

pileks committed:
fix order of config application in client constructor

pileks committed:
lint

dOrgJelli committed:
Merge pull request #1135 from polywrap/pileks-client-config-builder-methods

Client config builder methods

dOrgJelli committed:
fix: remove needless ignore paths

dOrgJelli committed:
fix: rename client-config to client-config-builder

dOrgJelli committed:
Merge branch 'pileks-client-config-builder-origin' of https://github.com/polywrap/monorepo into pileks-client-config-builder-origin

dOrgJelli commented:
Apologies for taking so long to review this @pileks, this was really well executed given how many potentially moving parts there are + legacy code. A few things I think we should remove in the next release (after this is out):
- `@polywrap/client-js`: `createPolywarpClient(...)` + `js/client/scripts/*`
- Move the ETH + IPFS + ENS defaults out of the `@polywrap/client-config-builder` package, and into their own "config bundle" package. Maybe call it `@polywrap/eth-ens-ipfs-config-js`.
    - Maybe have another "config bundle" package named `@polywrap/http-ws-config-js`. Not sure what to do about the FileSystem plugin and any others that are missed here...
    - Not even sure we need these "bundle packages". Could just move the "sane defaults" into their respective plugin packages `import { plugin, defaults } from "some-plugin"`
